db/db_with_timestamp_basic_test.cc

```diff
@@ -366,6 +366,60 @@ TEST_F(DBBasicTestWithTimestamp, UpdateFullHistoryTsLowWithPublicAPI) {
   s = db_->IncreaseFullHistoryTsLow(db_->DefaultColumnFamily(),
                                     ts_low_str_null);
   ASSERT_EQ(s, Status::InvalidArgument());
+  // test IncreaseFullHistoryTsLow concurrently. When actual ts_low is higher
+  // than requested ts_low, a try again error is returned.
+  DestroyAndReopen(options);
+  std::atomic<int> cnt{0};
+  const auto get_thread_id = [&cnt]() {
+    thread_local int thread_id{cnt++};
+    return thread_id;
+  };
+  SyncPoint::GetInstance()->DisableProcessing();
+  SyncPoint::GetInstance()->ClearAllCallBacks();
+  // what it takes to get try again
+  // write concurrently with another thread that is writing a bigger value
+  SyncPoint::GetInstance()->SetCallBack(
+      "VersionSet::LogAndApply:BeforeWriterWaiting", [&](void* /*arg*/) {
+        int thread_id = get_thread_id();
+        if (0 == thread_id) {
+          TEST_SYNC_POINT(
+              "DBBasicTestWithTimestamp::UpdateFullHistoryTsLowWithPublicAPI:"
+              "write_higher:0");
+          TEST_SYNC_POINT(
+              "DBBasicTestWithTimestamp::UpdateFullHistoryTsLowWithPublicAPI:"
+              "write_higher:1");
+        }
+      });
+  SyncPoint::GetInstance()->LoadDependency({
+      {"DBBasicTestWithTimestamp::UpdateFullHistoryTsLowWithPublicAPI:"
+       "write_higher:0",
+       "DBBasicTestWithTimestamp::UpdateFullHistoryTsLowWithPublicAPI:"
+       "WriteLowerStart"},
+      {"DBBasicTestWithTimestamp::UpdateFullHistoryTsLowWithPublicAPI:"
+       "WriteLowerStart",
+       "DBBasicTestWithTimestamp::UpdateFullHistoryTsLowWithPublicAPI:"
+       "write_higher:1"},
+  });
```

Comment:
I think it's trickier trying to set up test sync points dependencies between two threads that execute exactly the same set of functions. In this case, I wonder if we can use the following hack to get rid of `LoadDependency()` here.
1. In DBImpl::IncreaseFullHistoryTsLowImpl(), set up a sync point with callback and use the address of the local variable `edit` as argument. When sync points are hit by the two writer threads, we can establish a mapping from `std::this_thread::get_id()` to address of version edit.
2. In the callback registered with VersionSet::LogAndApply:BeforeWriterWaiting, we can figure out which thread hits this point first. For the first comer, we can access the `VersionEdit` via address and set its `full_history_ts_low` to be 25. For the thread that comes next, we set `full_history_ts_low` to 10.
3. For both threads, assert the results of execution.

Reply:
This is a good idea. Thank you for helping check this and proposing this change.

I tried this workaround out. It worked well to eliminate some edge cases, but there are still more work using test sync points to coordinate the two writing threads to guarantee a TryAgain error. We discussed offline and decided to simplify the test logic with a different workaround.